### ISSUE TITLE
roundto in assert statement not working as expected

### ISSUE THREAD

**Issue by bitti** [ledger/ledger#1925]
The sparsely documented `assert` directive might be sparsely tested as well. On my first try to use it I couldn't make it work when using `roundto`. Using a `floor` and a combination of multiplication, addition and division works but is awkward.

One easy way to reproduce:

```sh
$ ledger -f - <<LEDGER
assert floor(12.345 + 0.5) == 12
assert floor(12.345 * 10 + 0.5)/10 == 12.3
assert floor(12.345 * 100 + 0.5)/100 == 12.35
assert roundto(12.345, 0) == 12
assert roundto(12.345, 1) == 12.3
assert roundto(12.345, 2) == 12.35
LEDGER
```

This yields this output on my machine:
```
Ledger 3.2.1-20200518, the command-line accounting tool

Copyright (c) 2003-2019, John Wiegley.  All rights reserved.

This program is made available under the terms of the BSD Public License.
See LICENSE file included with the distribution for details and disclaimer.
While parsing file "", line 5:
Error: Assertion failed: roundto(12.345, 1) == 12.3
While parsing file "", line 6:
Error: Assertion failed: roundto(12.345, 2) == 12.35
```
So the first 4 assertions are fine, the last two are not. Are we supposed to use something else than `==` or is this a clear bug?

**Comment by tbm** [comment]
That's interesting. I added some debug statements and roundto() itself gets the right result:

```
    3ms  [DEBUG] We have an amount: 12.3
    3ms  [DEBUG] We have an amount: 12.35
```

@jwiegley any idea what the `assert` is complaining about?

**Comment by jwiegley** [comment]
Definitely a bug here. I changed one assert to:

```
assert roundto($12.345, 1) - $12.3 == 0
```

And here's the calculation output:

```
    1ms  [DEBUG] Calculating: roundto($12.345, 1) - $12.3 == 0
    1ms  [DEBUG]   ((roundto($12.345, 1) - $12.300) == 0) => ...
    1ms  [DEBUG] .  (roundto($12.345, 1) - $12.300) => ...
    1ms  [DEBUG] ..  roundto($12.345, 1) => ...
    1ms  [DEBUG] ...  $12.345 => ...
    1ms  [DEBUG] ...  $12.345 => $12.345
    1ms  [DEBUG] ...  1 => ...
    1ms  [DEBUG] ...  1 => 1
    2ms  [DEBUG] ..  roundto($12.300, 1) => $12.300
    2ms  [DEBUG] ..  $12.300 => ...
    2ms  [DEBUG] ..  $12.300 => $12.300
    2ms  [DEBUG] .  (roundto($12.300, 1) - $12.300) => $0.000
    2ms  [DEBUG] .  0 => ...
    2ms  [DEBUG] .  0 => 0
    2ms  [DEBUG]   ((roundto($12.300, 1) - $12.300) == 0) => false
```

This should work.

**Comment by bitti** [comment]
So is somebody working on this? Or should we have a more general discussion. Maybe this feature should just be removed, if nobody is using it and it isn't working as advertised?

**Comment by tbm** [comment]
@jwiegley can you look into this issue?

**Comment by maxnikulin** [comment]
I have realized that it is the same issue as #2329. `roundto` uses conversion from double internally and it causes huge power of 2 denominator in rational representation. See #2361 pull request for an attempt to address it.

**Comment by maxnikulin** [comment]
Sorry, I forgot to add "Closes" for this issue in the commit from #2361. It fixed
```
assert roundto(12.345, 1) == 12.3
```
As to
```
assert roundto(12.345, 2) == 12.35
```
12.34 (round to nearest even) vs. 12.35 is discussed in #1663.
